### ISSUE TITLE
SELFIE, media types, and content negotiation.

### ISSUE THREAD

**Issue by dblodgett-usgs** [opengeospatial/SELFIE#77]
The distinction between "meta" information and "data" information is defined by the context in which it is used (the use case that demands it). (more on that in #76)

Let's consider the general SELFIE happy path use case.

_**As a Web user, I want to find all the information available for an environmental feature, so I can find what I'm looking for and retrieve it.**_

As we've dissected this use case we've embraced HTTP-Range 14 (303 redirects), we've embraced OGC-API Features (html landing pages for features), and we've embraced schema.org JSON-LD in a `<script>` tag of a landing page. 

None of the above requires alternate media types. From the landing page, a human user or crawler with natural language tendencies is satisfied.

However, we recognize the potential to bring structure to data (including semantically enabled data) underlying the landing page content. I'll go one step further, SELFIE wouldn't be happening if we weren't layering this common 303->html+<script> pattern on top of a wide range of data systems that need such a discovery layer.

Given this, we need to make it clear that alternate media types and content negotiation, while not addressed specifically in SELFIE, are expected in a system that the SELFIE model is applied on top of.

Proposal: The findings of the IE will avoid specifying how to use content negotiation between the "meta" and "data" tiers. Standards for content negotiation by profile are emerging but we have not been able to evaluate them rigorously. As far as SELFIE will go is describing how to advertise that multiple content-types are available for a given URL.

**Comment by alpha-beta-soup** [comment]
> As far as SELFIE will go is describing how to advertise that multiple content-types are available for a given URL.

Given the necessity and centrality of an HTML landing page, a pattern the OAFeat API hearitly endorses, then I think we need to consider the appropriate [IANA link relation](https://www.iana.org/assignments/link-relations/link-relations.xhtml) to recommend if implementing this pattern. `alternate` seems simplest, but `describes`, `related`, `item` and perhaps others (`enclosure`?) seem possible. I'm only aware of a very small subset of these and I have certainly abused and ignored them in the past. In pygeoapi at present, outbound links to non-HTML (CSV, GeoJSON or RDF/JSON-LD, etc.) representations declared from the HTML view are simply `alternate` or `item`. Is that appropriate and adequate? This isn't necessarily a question of content negotiation, but of typing links in general. However, alongside the `rel` there's a `type`, presumably `application/ld+json` (but could be other kinds of RDF, if available)... so it does get back to conneg even if we're trying to avoid that discussion.

pygeoapi in practice:
- Link relations being included in the JSON representation: https://github.com/spatialdaotearoa/pygeoapi/blob/ld-json/pygeoapi/api.py#L479
- Link relations from the JSON representation hitting the HTML template in the `head`: https://github.com/spatialdaotearoa/pygeoapi/blob/ld-json/pygeoapi/templates/base.html#L16 
- The same link relations reaching the `body`: https://github.com/spatialdaotearoa/pygeoapi/blob/ld-json/pygeoapi/templates/collection.html#L33 (i.e. the "Links" part of this page: https://demo.pygeoapi.io/master/collections/obs)

**Comment by dblodgett-usgs** [comment]
Thanks @alpha-beta-soup I think this would be a good addition. When I wrote that I was thinking about alternate media types described in JSON-LD but I think it would be great to put the same in html. OK if we move to a new issue though? Could just copy your comment into a linked issue and change the comment here to recognize that issue?

**Comment by alpha-beta-soup** [comment]
@dblodgett-usgs you're welcome to organise the issues however you want; but if my comment doesn't fit here then I'm afraid I don't understand what this issue *is* about and it might need further clarification.

**Comment by dblodgett-usgs** [comment]
@alpha-beta-soup As described in this and the email and other issues from the last couple days, we seem to have developed a perception of their being two SELFIEs. One more webby, one more semantic. Many of the use cases involved in the IE are implementing sophisticated linked data systems with relatively simple range-14/OGC-API/HTML landing pages as a common thread. 

This issue is meant to say that what we _are_ doing is the HTML response. We will need to look to future work to specify the network architecture for other media types.

I do think that links we provide in JSON-LD (or HTML as you've offered here) ought to be annotated with the media-types they make available or link relations that describe the nature of the link. That is not network architecture though, that's content model.

In my mind, the content model is a different problem from the network architecture.  

Does that help? Comms about this stuff is too hard. :/

**Comment by denevers** [comment]
![December_Issue](https://user-images.githubusercontent.com/29091450/70804780-53159280-1d85-11ea-85b8-5f1319d2dfbb.png)

Here’s my mental representation of the discussion we had offline – up for discussion.  Sorry things sometime happens offline, sometimes, it’s better to have a conversation one to one when trying to narrow down the contentious issues.  It’s also to made snarky comments on the rest of you in private.
I don’t pretend it accurately represents what Alistair has in mind (or anyone else).

So, on this figure, we have one case of a governed architecture (like GSIP).   It show examples of what it can happen when a NIR is dereferenced.    All the option are “GSIP” compatible as far as I can tell, and even the provision proposed by Alistair (If I got it right) fit in the “GSIP” picture.
In recap, when a NIR is dereference, the client expects what we called a MIR, or a landing page. The MIR is a mixture of stuff about the thing itself and stuff about other representations and stuff for crawlers, etc.. of that thing.  We overly emphasised during the last few month on the “orange” part on the diagram (the whole subjectOf discussion).   GSIP is “Orange heavy”. This infrastructure is built to link various representations of real things that sits in different places.  It also provides crawlers with information it understand using crawler friendly ontologies (schema.org, and maybe we should consider other, like DCAT).  This information is for discovery and indexing.  The MIR can also have data  - the green part (link this thing to other thing) using a domain ontology (like ELFIE, GWML, GeoSciML).  We did not discuss this in Qc or very little.  
Let’s call this green data the canonical model for this architecture.  GSIP does not care of what’s in there.  It just transport it and expect client to traverse the links.  It become important for specialised client (one that understand ELFIE or GeoSciML).  GSIP just tells you where stuff is and what it is (format , conformance, etc.).  It also links to “non canonical” representations. The blue parts. 
In case A: the MIR has no canonical data (but *might* link to some).  Just links to representation.s We have GSIP cases for this when a list of representations can be inferred from a non-opaque URI.  GSIP has a mechanism to turn https://geoconnex.ca/id/waterwells/on.1104407 into a WFS request (or whatever) by reusing the id (1104407).  Everything on the landing page is inferred from the URI structure.  
Case B is the typical GSIP case.  Bit of data describing the real thing (using HyFeature in this case) and stuff for crawler and stuff for representations.  One of the representations MIGHT be a canonical representation but it does not have a special status.
Let’s skip C
D: is the other extreme (I would say it’s the BRGM example) when it’s all canonical data and what you get from the MIR is all what there is. No other representation, nothing for crawlers (or maybe a bit – Sylvain and Abdel might nudge this a bit and add a bit of orange for crawker, but still no “subjectOf” )
Now C is the point of discussion.  C is the same as B as far as GSIP is concerned. But some people (Alistair is not the only one, we also had this discussion early in GSIP) and dereferencing the NIR should in some circumstances lead directly to canonical data (so, directly to D), but recognising that some “orange” is required (for crawler and other non-canonical representations) – so it’s not “all D”.
How do we do this ?.  There are a couple of assumption Alistaire alluded to 
1-	 Assuming that crawler will just dereference the HTML version, the crawler stuff is really just needed in the HTML version (specialFormat). This “slightly” breaks the idea of conneg over format.  It means the HTML is a special case.  It also break GSIP a bit because it assume the only client needing the schema.org are browsers (eg. My QGIS app relies on RDF/TTL).  I personally don’t like it but I’m not outraged.
2-	Conneg by profile. If the client is “elfie” aware, in this particular infrastructure, it can ask for a particular profile.  This does not break GSIP, it’s something in parallel (same endpoint, different profile. If it’s supported, it can provide a specific content). It was one of the early GSIP discussion to have a “straightpipe” link to a representation, but we did not really explore “profile” at that point.  We excluded the “profile-less” straighpipe because it would be ambiguous and the developer would never know what to expect.
3-	“special” link in the MIR to identify the canonical data. In a nutshell, flag one of the representation.  We discuss having a type at some point, but it might also be something different from “subjectOf” (owl:sameAs – but sameAs is almost never what you want, it has way too many side effect. ).   The link should convey something like “this is where the rest of the data stuff is”

So.. this is my understanding

**Comment by dblodgett-usgs** [comment]
Thanks Eric. You are really digging into the complexity here.

In terms of what you've described here, my proposal is the SELFIE is focused on formalizing the use of NIRs for real world features in html representations of A and B (and I suppose D as well in some special limited cases). 

_What do you think about the proposal I put above?_ 

> Proposal: The findings of the IE will avoid specifying how to use content negotiation between the "meta" and "data" tiers. Standards for content negotiation by profile are emerging but we have not been able to evaluate them rigorously. As far as SELFIE will go is describing how to advertise that multiple content-types are available for a given URL.

I see this as leaving the door open to future experimentation and flexible implementations of linked data representations but doesn't put us in a position to try and write up and stand behind something we are only just starting to understand (at least I think we are starting to understand it).

**Comment by denevers** [comment]
Two concerns: 

- Is the semantic used in HTML has equivalent in RDF (I'm sure there is).  Otherwise, it makes HTML the only format to represent the link semantic/ model. 
- The HTML implementation discussed in this issue only addresses multiple formats, not multiple representation/provider/semantic (or does not disambiguate it).  It seems to work for one specific provider

The second item, as far as I understand, is required to address objectives in the activity plan

https://docs.google.com/document/d/1BG1XxDrkcf-JLHXordRzBIfLioNUOunUOuxeF5t-Eys/edit#heading=h.rn251ytbls77

**Objectives**
The objectives of the SELFIE include:

- Evaluate a proposed resource model for multi-provider environmental feature and observation registries
- Evaluate proposed HTTP behavior for non information resources and their representations
- Design and evaluate linked data feature information index resources with media-type, language, and profile content negotiation as an extension of the building blocks provided by WFS3.

I do understand though that the problem is way more complex than we originally estimated.

**Comment by dblodgett-usgs** [comment]
> Is the semantic used in HTML has equivalent in RDF (I'm sure there is). Otherwise, it makes HTML the only format to represent the link semantic/ model.

I don't see this as being the concern of what we are going to be able to draw conclusions about here here. If a given server/API wants to implement a variety of linked data graph "_views/shapes/etc." that's their prerogative. If one or more of them uses the same semantics as that used in the `<script>` section of the primary html landing page, that would be cool, but it's kind of irrelevant (if very interesting). I discussed this issue a bit over in #75 and want to be clear -- **I think we should document our findings but not draw conclusions on this part of the SELFIE scope.**

> The HTML implementation discussed in this issue only addresses multiple formats, not multiple representation/provider/semantic (or does not disambiguate it). It seems to work for one specific provider

I _am_ suggesting that we only draw conclusions about a pretty narrowly scoped piece of the puzzle, yes. That said, I think what I'm proposing we draw conclusions about is fundamental to the bigger picture of multiple representations, providers, or semantics. A HTML landing page with embedded semantic content targeted at search engines is level 0 of the more complex system.

> So, the HTML landing page you are referring to is the one for one of the representation (the landing page of a specific OGC API instance) - I think.
 
Yes and no. In the near term, it certainly could be as simple as one landing page to one GeoJSON feature with one NIR that can be used to link to it. But I'm imagining much much more down the road. I would like to get to a place where OGC-API is an API to a multi-provider, multi-representation, multi-semantic set of linked data skinned by a rich html landing page. But we can't just snap our fingers and have that happen. Getting NIRs in the first place is going to take some doing then we can start to fill in all the linked data content and figure out what the technical architecture really looks like.

---

I appreciate you pulling up those objectives and noting that the problem is way more complex than we imagined. (I also really appreciate having this dialogue with carefully worded comments in issue form, this will be a great resource for the ER!)

I think we can and should stick to that original plan. We've done a lot of "evaluating" and we've done a lot of "design". **What I'm saying here is that based on the evaluation and design, we need to hold back on coming to strong conclusions about the complex stuff but that we CAN make strong conclusions about some of the common and core stuff.**

**Comment by abhritchie** [comment]
Some random statements and observations ...
- A to D is a continuum with A and D being quite extreme endmembers. B and C are where the action is.
- D, 'pure' canonical data, is not in scope in the current suite of ELFIEs. These IEs have always been interested in matters of discovery and indexing, and the Spatial Data on the Web WG recommendations for the provision of data to support that. (Also, I have a gut feeling that D is handled if issues around B and C are addressed.)
- A is interesting. It's the library index card view of the world: some meta-data and an address of the thing (in (S)ELFIE's case a URL, not a Dewy Decimal System number). No content. An unlikely scenario in the _entry-level_ linked data web we are working to support, canonical data will always leak into the core representation, but it is useful is you have something like a feature served only by a WFS 2.0 server. (Again, I also have a gut feeling that A is handled if issues around B and C are addressed.)
- Where to start with B and C? First, the difference is very subtle to the point that I'm struggling to see it. But, it will be useful to explore what the special link and format are in your mind. Also, I'm interested to know how much of the orange-tinted view of the work is statements (with literal values) and how much is links.

**Comment by abhritchie** [comment]
Meanwhile. @denevers - can you clarify part of assumption 1 in the comment from December 14:
> It also break GSIP a bit because it assume the only client needing the schema.org are browsers (eg. My QGIS app relies on RDF/TTL).
Does GSIP break because GSIP assumes the only client needing schema.org are browsers, or does GSIP break because other aspects of GSIP (??) need schema.org content?

**Comment by denevers** [comment]
I can easily clarify: I'm wrong. 

The HTML page actually contains the RDF (in JSON-LD), so - does not break because GSIP can get a RDF (HTML is not a 1:1 representation)

**Comment by denevers** [comment]
The difference between B and C is that C has a special status for a canonical data representation that might not completely serialised in the landing page. For B, whatever is in the landing page is the canonical representation.  There might be more in one representation but it does not have a special status

**Comment by abhritchie** [comment]
I fear that Pandora's Can of Madness (TM) lies at the end of this path. How would you present this in the data?

**Comment by denevers** [comment]
I don't know.. I'm Hephaestus in this story

**Comment by abhritchie** [comment]
He makes things. Get your digital anvil out and make an example.

**Comment by denevers** [comment]
Need to consult the Oracle first (not Larry Ellison)

**Comment by dblodgett-usgs** [comment]
@abhritchie and @denevers -- can one of you summarize these comments into a disposition or question focused on the proposal here? I'll reiterate to try and bring some clarity to what I mean.

> Proposal: The findings of the IE will avoid specifying how to use content negotiation between the "meta" and "data" tiers. Standards for content negotiation by profile are emerging but we have not been able to evaluate them rigorously. As far as SELFIE will go is describing how to advertise that multiple content-types are available for a given URL.

>  I think we should document our findings but not draw conclusions on this part of the SELFIE scope.

>  What I'm saying here is that based on the evaluation and design, we need to hold back on coming to strong conclusions about the complex stuff but that we CAN make strong conclusions about some of the common and core stuff.

**Comment by abhritchie** [comment]
I've bashed this together for tomorrow's call. It's a straw-person for discussion, not a decree.
https://docs.google.com/document/d/1jNg2J1KR0BZ91Oi4iy-A0z-NuhyYwFy7x41gX-OaB-E/edit#

**Comment by ByronCinNZ** [comment]
Hi Alistair,

Do you think there might be something from this to contribute to OGC API Records work? (Formally CSW 4.)
I raised an issue to incorporate SELFIE findings in this work.

Cheers,
Byron

**Comment by denevers** [comment]
Thanks @abhritchie for this very useful document - it really helps when we have consistent vocabulary. 

I (@denevers , left hand on heart, right hand raised) fully agree on this sentence:

"A point of contention in SELFIE is how much of a description is provided (all the schema.org properties, all the domain properties, links to other representations, a mixture) and how. We’ve not explored this sufficiently to draw sound conclusions and @dblodgett-usgs has rightly asked us to step back here."

**Comment by abhritchie** [comment]
Hi @ByronCinNZ. I'm not familiar with the work so can't really comment. I'd hope (S)ELFIE would compliment most of the APIs.

**Comment by abhritchie** [comment]
@denevers: one hand on heart, the other raised. Does that mean anything these days?

**Comment by dblodgett-usgs** [comment]
I'm going to say we have agreed here and can move on to writing this up in the ER.

**Comment by dblodgett-usgs** [comment]
Content from #75, #76, #77, and others related to this has been copied and written up here. https://docs.google.com/document/d/1b-TdqWl3jf2jJD7dYvLkxTa6JEGxtUJAue7jSiaEE2s/edit#bookmark=id.jzbn3a8btrbf
